Improve 14/WAKU2-MESSAGE RFC format etc (#553)
vacp2p/rfc: content/docs/rfcs/14/README.md

---
slug: 14
title: 14/WAKU2-MESSAGE

> name: Waku v2 message RFC

Reviewer: Is the word `RFC` part of the template? I checked the template and it wasn't clear to me

Author: 🤷🏼

Reviewer 2: Seems odd to me, I'd cut it

status: stable

> category: Standards Track

Author: I am not sure if I picked the correct category for this RFC. Is this the right category for this RFC?

Reviewer: Where one can find descriptions of these categories?

Author: [Here](https://github.com/vacp2p/rfc/blob/master/content/docs/rfcs/template/README.md) you can find the template, but I cannot find the different categories description

Reviewer 2: think this is correct but defer to @kaiserd

Reviewer 3: I will add a category description to the COSS.

tags: waku/core-protocol
editor: Oskar Thorén <oskar@status.im>
contributors:
  - Sanaz Taheri <sanaz@status.im>
  - Aaryamann Challani <aaryamann@status.im>
  - Lorenzo Delgado <lorenzo@status.im>
---

# Abstract

> Waku v2 is a family of modular peer-to-peer protocols for secure communication. These protocols are designed to be secure, privacy-preserving, censorship-resistant, and able to run in resource-restricted environments. At a high level, Waku v2 implements a Pub/Sub messaging pattern over libp2p and adds capabilities to it.

Reviewer 4: Please use [semantic line breaks](https://sembr.org/)

> The present document specifies the Waku v2 message format and a way to encapsulate the messages sent with specific information security goals.

Reviewer: I'd rewrite it; for example, regarding this part: `messages sent with specific information security goals` should all messages have specific information security goals? and what are those goals?

Author: Sincerely, I am also confused about that phrase. I am not sure what was trying to say the original RFC author 👀

Reviewer 2: I'd also mention backwards compatibility as a goal

Specific information security goals means different messages have different requirements for e.g. confidentiality, authentication etc and this spec attempts to accommodate those in a modular fashion

# Motivation

When sending messages over Waku, there are multiple concerns:

- We may have a separate encryption layer as part of our application
- We may want to provide efficient routing for resource-restricted devices

> - We may want to provide compatibility with Waku v1 envelopes

Reviewer: Can we have a link to waku v1 envelope?

- We may want payloads to be encrypted by default

> - We may want to provide unlinkability for metadata protection

Reviewer: unlinkability `for` metadata protection? or unlinkability `and` metadata protection

Reviewer 2: provide message unlinkability... in order to get meta data protection

> This specification attempts to provide for these various requirements.

Reviewer: Are the above _concerns_ or _requirements_?

Reviewer: Regarding this comment:
the introductory sentence says:
`When sending messages over Waku, there are multiple **concerns**`
and the ending sentence says
`This specification attempts to provide for these various **requirements**.`
IMHO, these two terms mean different; now that we are updating the RFC, I'd suggest more consistent wording.

# Semantics

## Waku Message

> A waku message is constituted by the combination of data and attributes that, for example, a *publisher* sends to a *topic* and is eventually delivered to *subscribers*.

Reviewer: How `attribute` is distinguished from `data`?

Author: Do you want to specify a message attribute and what message data is?

I expect a reader to be knowledgeable about IT and messaging systems.

Reviewer: You have used these two terms, `data` and `attribute`, then the following subsection is titled `message attributes.`, so it comes across that the `Message data` section is missing. That is why I asked how you distinguish these two. Hope that clarifies it.

## Message Attributes

The `payload` attribute SHOULD contain the data to be sent.

The `contentTopic` attribute SHOULD specify a string identifier that can be used for content-based filtering.

> The `version` attribute MAY contain a version number to discriminate different types of payload encryption. If omitted, the value SHOULD be interpreted as version 0.

Reviewer: How about semantic line breaks?

Author: What do you mean by "semantic line breaks"? 👀 

I see two related and complementary phrases that do not require any line break for clarity/better readability.

Reviewer: I usually separate sentences after the period `.` i.e., put a line break, but not a big deal.

Reviewer 4: Semantic line breaks help mostly during the review process. At least each sentence should be in a new line, although I sometimes break a sentence as well, if there are several distinct clauses (this is up to you, though). Two newlines would indicate a new line in the rendered markdown. Remember that these single line breaks aren't rendered in markdown, so it's only really during reviews that they're visible.

> The `timestamp` attribute signifies the time at which the message was generated by its sender. This field SHOULD contain the Unix epoch time in nanoseconds. If the attribute is omitted, it SHOULD be interpreted as timestamp 0.

Reviewer 4: > If the attribute is omitted, it SHOULD be interpreted as timestamp 0.

I'm not sure why this clause is necessary? Couldn't it also be interpreted as `unset` to the same effect by some clients?

> The `ephemeral` attribute signifies the transient nature of the message. If this field is set to `true`, the message SHOULD be interpreted as ephemeral. Else, if the field is omitted or set to `false`, the message SHOULD be interpreted as non-ephemeral.

Reviewer 4: I understand that we probably don't want any coupling between this spec and the store specification, but I think it's OK to have an example here of what ephemeral MAY imply. Something like:
`For example, nodes interested in storing historical messages MAY choose to ignore ephemeral messages.`

# Wire Format

The `WakuMessage` wire format is specified using [protocol buffers v3](https://developers.google.com/protocol-buffers/).

```protobuf
syntax = "proto3";

message WakuMessage {
  bytes payload = 1;
  string content_topic = 2;
  uint32 version = 3;
  sint64 timestamp = 10;
  bool ephemeral = 31;
}

> ```

Author: **TO-DO:**

- [x] Add this protobuf to the Waku protocol buffers' repository

Author: PR opened here: https://github.com/vacp2p/waku/pull/1

Reviewer 2: Why? Shouldn't field names be under_score based on https://developers.google.com/protocol-buffers/docs/style?

Author: > Shouldn't field names be under_score based on [developers.google.com/protocol-buffers/docs/style](https://developers.google.com/protocol-buffers/docs/style)?

Yes, this is what I did here: `contentTopic` -> `content_topic`

See [13/WAKU2-STORE](/spec/13) for more details.
The `timestamp` attribute signifies the time at which the message was generated by its sender. This field SHOULD contain the Unix epoch time in nanoseconds. If the attribute is omitted, it SHOULD be interpreted as timestamp 0.

## Payloads
The `ephemeral` attribute signifies the transient nature of the message. If this field is set to `true`, the message SHOULD be interpreted as ephemeral. Else, if the field is omitted or set to `false`, the message SHOULD be interpreted as non-ephemeral.

Payloads are implemented using [protocol buffers v3](https://developers.google.com/protocol-buffers/).
# Wire Format

The `WakuMessage` wire format is specified using [protocol buffers v3](https://developers.google.com/protocol-buffers/).

```protobuf
syntax = "proto3";

message WakuMessage {
  bytes payload = 1;
  string contentTopic = 2;
  string content_topic = 2;
  uint32 version = 3;
  sint64 timestamp = 10;
  bool ephemeral = 31;
}
```

## Payload encryption

> Payload encryption depends on the `version` field.

Reviewer 2: Why is this gone?

Author: It is not gone, check line 82.

> The waku message payload MAY be encrypted. The message `version` attribute indicates the schema used to encrypt the payload data.

Reviewer: ```suggestion
The Waku message `payload` MAY be encrypted. The message `version` attribute indicates the schema used to encrypt the payload data.
```

Reviewer 2: I don't see how this rewrite makes it clearer (I mean original diff in PR), what's wrong with initial formulation?

Reviewer 2: Oh diff is confusing, this line is no longer referring to version 0?

> - **Version 0:**  The payload SHOULD be interpreted as unencrypted; additionally, it CAN indicate that the message payload has been encrypted at the application layer.

Reviewer: ```suggestion
- **Version 0:**  The `payload` SHOULD be interpreted as unencrypted; additionally, it CAN indicate that the message payload has been encrypted at the application layer.
```

> - **Version 0:**  The payload SHOULD be interpreted as unencrypted; additionally, it CAN indicate that the message payload has been encrypted at the application layer.

Reviewer: Why not `MUST`? i.e., `MUST` be interpreted?

Author: Same question for you, why not `SHOULD`? Can you elaborate?

Reviewer: Yes, sure; for reference, based on https://rfc.vac.dev/spec/1/#language and the document referred therein https://www.rfc-editor.org/rfc/rfc2119, the term `SHOULD` indicates _recommendation_ whereas `MUST` indicates _an absolute requirement_. So in here, when the version is 0 (or 1), to me, the payload is absolutely unencrypted  (or encrypted), i.e., there is no doubt. Hence the term `MUST` looks more appropriate.

Reviewer 4: Not sure I agree, @staheri14. When version is `0` it simply means that Waku is not enforcing some encryption scheme. Applications are free to encrypt the payload as they wish (and most do). `SHOULD` is therefore more appropriate, IMO.

Reviewer: >  Applications are free to encrypt the payload as they wish (and most do). SHOULD is therefore more appropriate, IMO.

I see what you mean, right, for version 0 it makes sense to use `SHOULD`. But in the other cases i.e., version value of 1 and 2, I think that needs to be updated to `MUST` following my previous reasoning.

Reviewer: Or, are we going to argue that if the version is e.g., `1`, the application still can decide to use a different encryption scheme or none at all? @jm-clius if that is the case, then `SHOULD` is appropriate.

Reviewer 4: Indeed. I don't think from an RFC perspective we can force applications to use `version` in a certain way, although it's clearly recommended.

> - **Version 0:**  The payload SHOULD be interpreted as unencrypted; additionally, it CAN indicate that the message payload has been encrypted at the application layer.

Reviewer 4: ```suggestion
- **Version 0:**  The payload SHOULD be interpreted as unencrypted; additionally, it MAY indicate that the message payload has been encrypted at the application layer.
```

`CAN` is not a [recognised RFC keyword](https://rfc.vac.dev/spec/1/#language).

> - **Version 2:** The payload SHOULD be encrypted using [35/WAKU2-NOISE](/spec/35).

Reviewer: Ditto about `SHOULD` vs `MUST` here

> Any `version` value not included in the list is reserved for future use, and the payload SHOULD be interpreted as unencrypted.

Reviewer 4: Something like `...unencrypted by the Waku layer.` We don't know anything about whether it's encrypted or not.

> ## Confidentiality, integrity, and authenticity

Reviewer 2: I don't understand what is going on here in this diff or what the goal of rewrite is

Author: Conforming to the RFC template sections. The GH UI diff is not "understanding" the changes.

- **Version 2:** The payload SHOULD be encrypted using [35/WAKU2-NOISE](/spec/35).
Waku Noise protocol provides symmetric encryption and asymmetric key exchange.

This indicates that payloads MUST be encrypted using [WAKU2-PAYLOAD](/spec/26).
Any `version` value not included in the list is reserved for future use, and the payload SHOULD be interpreted as unencrypted.

This provides for asymmetric and symmetric encryption.
Key agreement is out of band.
It also provides an encrypted signature and padding for some form of unlinkability.
# Security Considerations

### Version 2
## Confidentiality, integrity, and authenticity

This indicates that payloads MUST be encoded using [35/WAKU-NOISE](/spec/35).
The level of confidentiality, integrity, and authenticity of the waku message payload is discretionary. Accordingly, the application layer shall utilize the encryption and signature schemes supported by Waku v2 to meet the application-specific privacy needs.

This provides for symmetric encryption and
asymmetric key-exchange protocols.
## Reliability of the `timestamp` attribute

> # Differences from Whisper / Waku v1 envelopes

Reviewer: Is this intentional to remove this section? curious to know why

Author: IMHO, there is little value in comparing the Waku v2 message format with whisper/waku v1 envelope formats for this RFC. But it may be relevant in another RFC like the waku v1 to waku v2 bridging RFC.

We can restore it if we think it adds some value. Otherwise, I'd rather keep the RFC succinct.

Author: @oskarth How relevant is this "Whisper/Waku V1 vs Waku v2 comparison" section?

> For example, a malicious actor can arbitrarily set the `timestamp` of a waku message to a high value so that it always shows up as the most recent message in a chat application. Applications using waku messages’ `timestamp` attribute are recommended to use additional methods for more robust message ordering. An example of how to deal with message ordering against adversarial message timestamps can be found in the Status protocol, see [6/PAYLOADS](https://specs.status.im/spec/6#clock-vs-timestamp-and-message-ordering).

Reviewer 4: Semantic breaks would help here to review each line separately.

> For example, a malicious actor can arbitrarily set the `timestamp` of a waku message to a high value so that it always shows up as the most recent message in a chat application. Applications using waku messages’ `timestamp` attribute are recommended to use additional methods for more robust message ordering. An example of how to deal with message ordering against adversarial message timestamps can be found in the Status protocol, see [6/PAYLOADS](https://specs.status.im/spec/6#clock-vs-timestamp-and-message-ordering).

Reviewer 4: I think we need to settle on `Waku` as always being capitalised. @kaiserd, thoughts?

> The waku message `ephemeral` attribute is set by the sender. Since there is currently no incentive mechanism for network participants to behave correctly, this field is inherently insecure. A malicious actor can tamper with the value of a waku message’s `ephemeral` attribute, and the receiver would not be able to verify the integrity of the message.

Reviewer 4: Decoupling from existing specs makes sense to me (i.e. no direct reference to relay/store specs), but I think an example is necessary to clarify. Could just be "for example, nodes storing historical messages, may ignore messages..." etc.

Since Waku v2 is using libp2p PubSub, some of these fields can be dropped.
The previous `topic` field corresponds to `contentTopic`.
The previous `data` field corresponds to the `payload` field.
## Reliability of the `ephemeral` attribute

# Security Consideration
The waku message `ephemeral` attribute is set by the sender. Since there is currently no incentive mechanism for network participants to behave correctly, this field is inherently insecure. A malicious actor can tamper with the value of a waku message’s `ephemeral` attribute, and the receiver would not be able to verify the integrity of the message.

## Confidentiality, integrity, and authenticity 
It is up to the application layer as to what level confidentiality, integrity and authenticity of the `payload` of `WakuMessage` matters. 
Accordingly, the application layer shall utilize the encryption and signature schemes supported in WAKU2 to meet the application-specific privacy needs.
The set of supported schemes in WAKU2 is presented in [WAKU2-PAYLOAD](/spec/26).
## Copyright

> ## Reliability of the WakuMessage timestamp

Reviewer 2: Why is this removed

Author: It is not removed, but with the format changes, it seems that the diff is not detecting it. Check line 113.

> TBD

Author: I am unsure exactly what to put here 🤔

Reviewer: I'd say the links you have already put in the text.

Reviewer 2: It should be summarized links that are in text

---
slug: 14
title: 14/WAKU2-MESSAGE
name: Waku v2 message

> status: draft

Author: I put it back to `draft` as per @oskarth comment. We can change to `stable` in a different PR.

Waku message attributes are key-value pairs of metadata associated with a message. 
And the message data payload is the part of the transmitted Waku message that is the actual message information.
The data payload is also treated as a Waku message attribute for convenience.

## Message Attributes

* The `payload` attribute MUST contain the message data payload to be sent.

* The `content_topic` attribute MUST specify a string identifier that can be used for content-based filtering.

* The `version` attribute, if present, contains a version number to discriminate different types of payload encryption.
If omitted, the value SHOULD be interpreted as version 0.

* The `timestamp` attribute, if present, signifies the time at which the message was generated by its sender.
This attribute MAY contain the Unix epoch time in nanoseconds.
If the attribute is omitted, it SHOULD be interpreted as timestamp 0.

* The `ephemeral` attribute, if present, signifies the transient nature of the message.
For example, an ephemeral message SHOULD not be persisted by the Waku network.
If this attribute is set to `true`, the message SHOULD be interpreted as ephemeral.

> Else, if the attribute is omitted or set to `false`, the message SHOULD be interpreted as non-ephemeral.

Reviewer 5: ```suggestion
If, instead, the attribute is omitted or set to `false`, the message SHOULD be interpreted as non-ephemeral.
```

# Wire Format

The Waku message wire format is specified using [protocol buffers v3](https://developers.google.com/protocol-buffers/).

```protobuf
syntax = "proto3";

message WakuMessage {
  bytes payload = 1;
  string content_topic = 2;
  optional uint32 version = 3;
  optional sint64 timestamp = 10;
  optional bool ephemeral = 31;
}
```

# Payload encryption

The Waku message payload MAY be encrypted.
The message `version` attribute indicates the schema used to encrypt the payload data.

- **Version 0:**
  The payload SHOULD be interpreted as unencrypted; additionally, it CAN indicate that the message payload has been encrypted at the application layer.

- **Version 1:**

> The payload SHOULD be encrypted using Waku v1 payload encryption. [26/WAKU-PAYLOAD](/spec/26).

Reviewer 5: ```suggestion
The payload SHOULD be encrypted using Waku v1 payload encryption specified in [26/WAKU-PAYLOAD](/spec/26).
```

This provides asymmetric and symmetric encryption.
The key agreement is performed out of band.
And provides an encrypted signature and padding for some form of unlinkability.

- **Version 2:**

> The payload SHOULD be encrypted using [35/WAKU2-NOISE](/spec/35).

Reviewer 5: Waku Noise RFC specifies how to do different things (key exchange/encryption/etc). I'm not sure we should talk exclusively about encryption, although of course it's main use. E.g. in a key exchange the first messages are not "encrypted" but ultimately will.

```suggestion
The payload SHOULD be encoded according to [35/WAKU2-NOISE](/spec/35).
```

Waku Noise protocol provides symmetric encryption and asymmetric key exchange.

> Any `version` value not included in the list is reserved for future use, and the payload SHOULD be interpreted as unencrypted by the Waku layer.

Reviewer 5: ```suggestion
Any `version` value not included in this list is reserved for future specification: when a `version` values not specified by this document is employed, the payload SHOULD be interpreted as unencrypted by the Waku layer.
```
